custom/src/main/java/co/elastic/otel/config/DynamicInstrumentation.java

```diff
@@ -0,0 +1,261 @@
+                Field tracerEnabledField = sdkTracer.getClass().getDeclaredField("tracerEnabled");
+                tracerEnabledField.setAccessible(true);
+                // Update is synced but the reader is NOT necessarily so this is eventual
+                // consistency, takes effect when the application passes a sync boundary
+                synchronized (sdkTracer) {
+                  tracerEnabledField.set(sdkTracer, tConfig.isEnabled());
+                }
+                if (logger.isLoggable(Level.CONFIG)) {
+                  String name = instrumentationScopeInfo.getName();
+                  if (name.startsWith(INSTRUMENTATION_NAME_PREPEND)) {
+                    name = name.substring(INSTRUMENTATION_NAME_PREPEND.length());
+                  }
+                  activatedTracers.add(name);
+                  activatedTracers.add(tConfig.isEnabled() ? "enabled" : "disabled");
+                }
+              } catch (NoSuchFieldException | IllegalAccessException e) {
+                throw new RuntimeException(e);
+              }
+            });
+    if (logger.isLoggable(Level.CONFIG)) {
+      logger.log(Level.CONFIG, "Activated Tracers: " + activatedTracers);
+    }
+  }
+
+  public static void reenableTracesFor(String instrumentationName) {
+    UpdatableConfigurator.INSTANCE.put(
+        InstrumentationScopeInfo.create(INSTRUMENTATION_NAME_PREPEND + instrumentationName),
+        TracerConfig.enabled());
+    updateTracerConfigurations(GlobalOpenTelemetry.getTracerProvider());
+  }
+
+  public static void disableTracesFor(String instrumentationName) {
+    UpdatableConfigurator.INSTANCE.put(
+        InstrumentationScopeInfo.create(INSTRUMENTATION_NAME_PREPEND + instrumentationName),
+        TracerConfig.disabled());
+    updateTracerConfigurations(GlobalOpenTelemetry.getTracerProvider());
+  }
+
+  public static class UpdatableConfigurator implements ScopeConfigurator<TracerConfig> {
+    public static final UpdatableConfigurator INSTANCE = new UpdatableConfigurator();
+    private final ConcurrentMap<String, TracerConfig> map = new ConcurrentHashMap<>();
+
+    private UpdatableConfigurator() {}
+
+    @Override
+    public TracerConfig apply(InstrumentationScopeInfo scopeInfo) {
+      return map.getOrDefault(scopeInfo.getName(), TracerConfig.defaultConfig());
+    }
+
+    public void put(InstrumentationScopeInfo scope, TracerConfig tracerConfig) {
+      map.put(scope.getName(), tracerConfig);
+    }
+  }
+
+  static {
+    if ("true".equals(System.getProperty(INSTRUMENTATION_DISABLE_OPTION + ".checker"))
+        || "true".equals(System.getenv("ELASTIC_OTEL_JAVA_DISABLE_INSTRUMENTATIONS_CHECKER"))) {
+      new Thread(new OptionChecker(), "Elastic dynamic_instrumentation checker").start();
+    }
+  }
+
+  static class OptionChecker implements Runnable {
+    private Map<String, Boolean> alreadyDisabled = new HashMap<>();
```

Comment:
This algorithm is stateful due to carrying over `alreadyDisabled` over the loop iterations.

I think this can be greatly simplified because we have that state already in another place, namely `UpdatableConfigurator.map`:

What I would do is something like adding a `boolean setOnlyDisabled(Set<String> scopeNames)` method to `UpdateableConfigurator` which:
 * Reenables all `TraceConfigs` in `map` which are not mentioned in `scopeNames`
 * Makes all elements in `scopeNames` disabled in `map`
 
 As a result, you can now just pass the the desired target state (the list of disabled instrumentations coming form the config) to it periodically and dependent on the return value update the tracers.

Reply:
That would make it beautifully simple! However that assumes only this checker (ie only this property) sets disablement, which is not the case. Example scenario: central config disables instrument X. `disable_instrumentations` property remains empty. This thread executes - result would be that central config client would disable X and then this thread would re-enable it almost immediately.

Reply:
But that does identify that there is a priority conflict in which decides enablement if both are used for the same instrumentation. As this thread is really only here for testing the functionality (although it can be used in production), I'll leave it as is since deciding priority will make it more complicated. I'll add a note about that though

Reply:
> Example scenario: central config disables instrument X. disable_instrumentations property remains empty

I was expecting that the config merge happens beforehand and that this thread operates on the already merged conflicts and therefore is the only one performing the updates. However, as this is not the case you are totally right.